streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java

```diff
@@ -200,6 +201,7 @@ synchronized void evict() {
         }
         remove(eldest);
         cache.remove(eldest.key);
+        dirtyKeys.remove(eldest.key);
```

Comment:
Makes sense, thanks!

Have two other thoughts about the PR:

1. Instead of trying to add `dirtyKeys.remove(eldest.key);`, would it be better to consider checking if 

```
!value.isDirty() && dirtyKeys.contains(key)
```

in the `synchronized void put(final Bytes key, final LRUCacheEntry value)`, so that we would not hide the potential issue. And if it did happen, we could either throw a more meaningful exception or log warning and delete it manually if we want to proceed it to unblock users (personally I'd prefer the first option to find out the issue asap which may delay the unblocking but worthwhile).

2. Another random thought about the root cause -- since we do not have more information besides the stack trace I can just keep guessing -- is that, in `synchronized void flush() {`:

```
        for (Bytes key : dirtyKeys) {
            final LRUNode node = getInternal(key);
            if (node == null) {
                throw new IllegalStateException("Key found in dirty key set, but entry is null");
            }
            entries.add(new ThreadCache.DirtyEntry(key, node.entry.value, node.entry));
            node.entry.markClean();
        }
        listener.apply(entries);
        dirtyKeys.clear();
```

Let's say we have completed the forloop, so that entries' `isDirty` bit is already reset, and then we call `listener.apply`, which in turn results in lots of computations that cause us to call flush (so we can calling flush again inside the apply function of this flush) on the same NamedCache, then is it likely that the entry is gone but the dirtyKeys is not cleared yet?